File: app/src/main/java/com/sdex/activityrunner/intent/LaunchParamsExtraListAdapter.java
package com.sdex.activityrunner.intent;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.sdex.activityrunner.R;
import java.util.ArrayList;
import java.util.List;

public class LaunchParamsExtraListAdapter extends
  RecyclerView.Adapter<LaunchParamsExtraListAdapter.ViewHolder> {

  private final Callback callback;
  private List<LaunchParamsExtra> items = new ArrayList<>();

  public LaunchParamsExtraListAdapter(Callback callback) {
    this.callback = callback;
  }

  @NonNull
  @Override
  public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
    final View view = inflater.inflate(R.layout.item_launch_param_extra, parent, false);
    return new ViewHolder(view);
  }

  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
    final LaunchParamsExtra item = items.get(position);
    holder.bind(item, callback);
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  public void setItems(List<LaunchParamsExtra> items) {
    this.items.clear();
    this.items.addAll(items);
    notifyDataSetChanged();
  }

  public interface Callback {

    void onItemSelected(int position);
  }

  public static class ViewHolder extends RecyclerView.ViewHolder {

    public final TextView keyView;
    public final TextView valueView;

    public ViewHolder(View itemView) {
      super(itemView);
      this.keyView = itemView.findViewById(R.id.key);
      this.valueView = itemView.findViewById(R.id.value);
    }

    public void bind(LaunchParamsExtra item, Callback callback) {
      keyView.setText(item.getKey());
      valueView.setText(item.getValue());
      itemView.setOnClickListener(v -> callback.onItemSelected(getAdapterPosition()));
    }
  }
}

File: app/src/main/java/com/sdex/activityrunner/intent/LaunchParamsExtraListAdapter.kt
package com.sdex.activityrunner.intent

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.sdex.activityrunner.databinding.ItemLaunchParamExtraBinding

class LaunchParamsExtraListAdapter :
    RecyclerView.Adapter<LaunchParamsExtraListAdapter.ViewHolder>() {

    private var items: List<LaunchParamsExtra> = ArrayList()
    private var viewMode = false
    var callback: Callback? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(ItemLaunchParamExtraBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = items[position]
        holder.bind(item, callback, viewMode)
    }

    override fun getItemCount(): Int = items.size

    @SuppressLint("NotifyDataSetChanged")
    fun setItems(items: List<LaunchParamsExtra>, viewMode: Boolean = false) {
        this.items = items
        this.viewMode = viewMode
        notifyDataSetChanged()
    }

    interface Callback {

        fun onItemSelected(position: Int)

        fun removeItem(position: Int)
    }

    class ViewHolder(
        private val binding: ItemLaunchParamExtraBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(item: LaunchParamsExtra, callback: Callback?, viewMode: Boolean) {
            binding.key.text = item.key
            binding.value.text = item.value
            binding.remove.isVisible = !viewMode
            if (viewMode) {
                binding.root.setOnClickListener(null)
            } else {
                binding.root.setOnClickListener {
                    callback?.onItemSelected(bindingAdapterPosition)
                }
                binding.remove.setOnClickListener {
                    callback?.removeItem(bindingAdapterPosition)
                }
            }
        }
    }
}
